nixos/modules/services/mail/dovecot.nix

```
@@ -1,19 +1,91 @@
 { config, lib, pkgs, ... }:
 
 let
-  inherit (lib) attrValues concatMapStringsSep concatStrings
-    concatStringsSep flatten imap1 literalExpression mapAttrsToList
-    mkEnableOption mkIf mkOption mkRemovedOptionModule optional optionalAttrs
-    optionalString singleton types mkRenamedOptionModule nameValuePair
-    mapAttrs' listToAttrs filter;
-  inherit (lib.strings) match;
+  inherit (lib) attrValues concatMapStringsSep concatStrings concatStringsSep
+    flatten imap1 isBool isDerivation isInt isList isPath isString
+    literalExpression mapAttrsToList mkEnableOption mkIf mkMerge mkOption
+    mkRemovedOptionModule optional optionalAttrs optionalString singleton
+    types mkRenamedOptionModule nameValuePair mapAttrs' listToAttrs filter
+    filterAttrs;
+  inherit (lib.generators) toPretty;
+  inherit (lib.lists) findFirstIndex length sublist reverseList;
+  inherit (lib.strings) escape match splitString;
 
   cfg = config.services.dovecot2;
   dovecotPkg = pkgs.dovecot;
 
   baseDir = "/run/dovecot2";
   stateDir = "/var/lib/dovecot";
 
+  /* Helper: Format string in way that may strech over multiple
+   * lines in a Dovecot configuration file
+   *
+   * This is a little complicated since Dovecot does not like when the last
+   * line is empty, so we first have to truncate all trailing empty lines
+   * before when can join them using the continuation line syntax.
+   */
+  formatDovecotMultiline =
+    string:
+      let
+        # Split input into line list
+        lines = splitString "\n" string;
+        # Find index of last non-empty line in line list
+        lastContentIndex = length lines - (findFirstIndex (
+          line: match "[[:space:]]*" line == null  # <=> Line is not empty
+        ) (length lines) (reverseList lines));
+        # Drop empty trailing lines from line list
+        trimmedLines = sublist 0 lastContentIndex lines;
+      in
+        # Reassemble content without trailing lines but with continuation markers
+        concatStringsSep " \\\n" trimmedLines;
+
+  /* Helper: Apply string escaping and quotation to Dovecot string if needed
+   *
+   * Dovecot is sometimes touchy about strings being quoted, so this only
+   * quotes the string when it is actually needed.
+   */
+  escapeDovecotString = string:
+    let
+      escapedString = escape ["\\" "\""] string;
+    in   if match ''[[:space:]]*<.*|.*[[:space:]#"\].*|'' string != null
```

```suggestion
    in   if match ''[[:space:]]*<.*|.*[[:space:]].*|'' string != null
```

Actually, this regexp is invalid, I didn't check that before. Is this maybe what it should be?

---

No, the regexp works just fine.

The following Nix configuration:

```nix
    pluginSettings = {
         test_quote = "\"";
         test_comment = "a#b";
         test_space = " ";
         test_backslash = "\\";
         test_unquoted = "abc";
    };
```

Results, using the submitted regexp, in the following Dovecot configuration:

```
plugin {
    test_backslash = "\\"
    test_comment = "a#b"
    test_quote = "\""
    test_space = ""
    test_unquoted = abc
}
```

The condition `if match ''[[:space:]]*<.*|.*[[:space:]#"\\].*|'' string != null` also appears to work.

---

Oh, I guess this hits #1537. It fails to evaluate on my Mac.

```
error: invalid regular expression '[[:space:]]*<.*|.*[[:space:]#"\].*|'
```

The | needs to be escaped for it to work everywhere:

```suggestion
    in   if match ''[[:space:]]*<.*\|.*[[:space:]#"\].*\|'' string != null
```

---

I checked the linked issue but the mentioned problem does not appear to be present inside a Darling (macOS on Linux emulator) container, so I’m unable to reproduce this macOS-only behaviour locally. The pattern is supposed to say “starts with `<` (ignoring whitespace) OR contains at least one whitespace, `#`, `"` or `\` character OR is empty” in any case.

I’ve updated the regexp to something that should hopefully work on macOS too – if it’s just the empty variant it dislikes. Please also check if the test vector I posted produces the same results: https://github.com/NixOS/nixpkgs/pull/286184#discussion_r1520494131 (there actually was a bug that single-lined strings containing only whitespace would also be trimmed).

---

Great, looks like that works! The test cases give the same output here.

I thought maybe you could get the same behavior by overriding nix to use clangStdenv but apparently not. Well, doesn't matter now anyway.